### ISSUE TITLE
Multiple requests to synchronous views are handled sequentially, not concurrently. Django 3.1.3, Channels 3.0.2, Daphne 3.0.1.

### ISSUE THREAD

**Issue by StefanCardnell** [django/channels#1587]
`Pipfile`:

```

[packages]
Django = "==3.1.3"
asgiref = "==3.2.10"
channels = "==3.0.2"
daphne = "==3.0.1"
```

(Note: asgiref 3.3.10 was also tried)

***Code / Minimal example***

I started a completely new Django + Django channels project to verify I can still see the same behaviour.

`views.py`:

```py

def view_1(request):
    for i in range(10):
        time.sleep(1)
        print(request, i)

    return HttpResponse(b'Hello World #1!')


def view_2(request):
    return HttpResponse(b'Hello World #2!')
```

`asgi.py`:

```py

import os

from channels.routing import ProtocolTypeRouter, URLRouter
from django.core.asgi import get_asgi_application
from django.urls import re_path

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'channels_test.settings')

application = ProtocolTypeRouter({
    'http': URLRouter([
        re_path('', get_asgi_application())
    ])
})
```

`urls.py`:

```py

urlpatterns = [
    path('admin/', admin.site.urls),
    path('view1/', views.view_1),
    path('view2/', views.view_2),
]
```

I have two Django views. `View1` and `View2`. Both are synchronous. As you can see, `View1` sleeps for 10 seconds before returning a response. Calls to View2 respond instantaneously.

**How you're running Channels (runserver? daphne/runworker? Nginx/Apache in front?)**

Issue is seen in both runserver and also when running a single daphne worker. Nginx is serving the requests.

***What Happened***

If I open up a browser and go to localhost/view1, I can see in the console the printouts of the sleep for 10 seconds. If, while view1 is still processing, I go to localhost/view2, the page does **not** respond until view1 has finished processing. 

***What I expected***

I should not expect View2 to have to wait for View1 to finish before returning a response, the requests should run concurrently. If I revert back to channels 2.4.0 with the following pip setup:

`Pipfile`:

```

[packages]
Django = "==3.1.3"
asgiref = "==3.2.10"
channels = "==2.4.0"
daphne = "==2.5.0"
```

and the following `asgi.py`:

```py

import os

from channels.routing import ProtocolTypeRouter, URLRouter
from channels.routing import AsgiHandler
from django.urls import re_path

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'channels_test.settings')

import django
django.setup()
application = ProtocolTypeRouter({
    'http': URLRouter([
        re_path('', AsgiHandler)
    ])
})

```

I no longer see the behaviour of the ticket. As I did not expect a regression in behaviour for synchronous views after upgrading django / channels, I expected to see this same older behaviour after upgrading channels. 

***Workarounds***

1) Convert the synchronous views to asynchronous views. This is not feasible for me as my entire project uses django rest framework which does not support asynchronous views (as Django does not provide async generic views).

2) Reverting back to Channels 2.X / Daphne 2.X. 

***Investigation / Notes***

- While I know Django introduced async views, I did not expect a regression in behaviour for existing uses of synchronous views with ASGI. I did not see documentation indicating this behaviour should be expected.

- For my project, all APIs sent are now processed sequentially resulting in a very unresponsive front-end.

- I am certain this is down to the use of Django's `ASGIHandler` and its use of `thread_sensitive=True` in `sync_to_async`. Synchronous views are all placed in the same thread with `thread_sensitive=True`, so I'm assuming this is the cause of the issue - and wasn't an issue with channel's own version of `ASGIHandler`. If I access `django.core.handlers.base` and modify `thread_sensitive` of `_get_response_async` to `False`, the view responds instantly, like normal:

`Line 228` of `django.core.handlers.base`

```py

            if not asyncio.iscoroutinefunction(wrapped_callback):
                wrapped_callback = sync_to_async(wrapped_callback, thread_sensitive=False)

```

- I wasn't sure whether to report as a Channels or a Django issue. As using Django's `get_asgi_application` is part of the recommended way of upgrading channels from `2.X` to `3.X`, and this is all part of the Django ecosystem, it felt more appropriate here.

**Comment by carltongibson** [comment]
Hi @StefanCardnell — thanks for the good report. Have a look at #1582, which touches the same topic. 

We'll need to adjust the handling here but (for lack of bandwidth) I'm yet to be able to sit down with it and have a think about exactly how, so all input welcome.

**Comment by StefanCardnell** [comment]
Hi @carltongibson. Thanks for the response. The PR solves effectively the same issue caused by thread sensitivity, just originating from a different place - an asgiref update. The PR will fix Channel's `ASGIHandler` so Channels 2.X work with asgiref > 3.3.

But obviously `ASGIHanlder` has been deprecated since channels 3.X and favour of swapping to usage of Django's own `ASGIHandler` with its own `thread_sensitive=True`. That being the root cause in this case.

It could be a core issue in Django's new ASGI API but as Channels uses the ASGI API and performs a few monkey patches here and there, again, not really sure where it should be fixed. Either way, unfortunately a show-stopper for us being able to upgrade to the latest channels.

**Comment by carltongibson** [comment]
Hey @StefanCardnell — I'd suggest that in the short-run at least (unless you're actively using Django `async def` views already) that you route the Django views via a WSGI app and then run ASGI separately for Channels related bits. 

Clearly that's not the end goal, but it keeps the well battle tested parts (the WSGI pipeline) in play, and should let you progress. 

This is something I will have a proper think about over Christmas.

**Comment by jamiegau** [comment]
This may be related to this problem.
I reported this a while back but it was said to be not specific enough.
I am trying to upgrade an app to the latest version and when I did, I started having my app lockup on quires to the django server.   If the "channels", is added to INSTALLED_APPS, and I then try and do multiple requests, in my case, those requests never return and chrome reports them as "pending" forever.
The Django logs indicate the requests have completed..  they just never to on the browsers end.

Problems go away when I drop "channels" from INSTALLED_APPS, but, I need channels, so I am stuck.
channels==3.0.2
Django==3.1.4
djangorestframework==3.12.2

I am using async implementation in channels.

This is quite a show stopper.

**Comment by carltongibson** [comment]
Hi @jamiegau This doesn't seem like the same issue. The problem here is that requests are being handled sequentially, not that they're not appearing in the browser. If you can provide more detailed information maybe we can look on the original issue, but I'd like to avoid clouding the issue here. Thanks!

**Comment by jamiegau** [comment]
@carltongibson It has very similar issues.  In that, only when 'channels' is activated in a project, when trying to do concurrent requests to django, some type of race condition occurs and all the concurrent calls never return, resulting in the browners debugger indicating those requests are pending.. and never return or even error until I kill the django server.  The original issue may indicate they are blocked as so only occur sequentially when this occurs, while I get a complete race condition lockup.

Sounds like the same code segment to me.  And if devs are going to be looking into this, they should be aware of other issues exposed by this as so they keep that in mind when fixing the other cause too.  Personally I think they must be associated as they are both based on concurrency issues.

**Comment by JhonFrederick** [comment]
I have  the same problem. I run the project and if I have two browser tabs and on tab 1 the request takes a long time, on tab 2 any request I make must wait until tab 1 the request is completed.
I am using Django channels and Django but I discovered that the problem doesn't happen when I run my project with Ubuntu (Subsystem for Windows) but if when I run the project with Windows. In both cases I changed  the python version (python3.7 and then python3.8), the packages version and nothing. So I tried only execute my project using daphne and,  Which was my sorprise? with or without the channels package the problem appears. On the dev team some have the problem on both windows and ubuntu but in the case of them changing the python version (python 3.7) solve the problem.

```
Original Packages   ------- >  After Upgrade
[packages]                       [packages]                             
Django==2.2.10                  Django==3.1.5
channels==2.4.0                 channels==3.0.3
daphne==2.5.0                   daphne==3.0.1
asgiref==3.2.10                  asgiref==3.3.1
....                            ....
```
 It should be emphasised that it did not happen before.

**Comment by ivanmviveros** [comment]
I have the same problem as part of the same dev team of @JhonFrederick , in addition to their description for the issue, another strange behaivor is that when I tried multiple long request in different tabs on the same browser, the requests runs sequentially, but if I try the same requests over distintc browsers, the requests execution is as expected, this issue is a critical block in our system because the long requests which I mean are related to an external webservice and wait for their result, but the result itself is only returned after the webservice send some post requests to our Django system, then the external webservice can never accomplish the process because the server rejects the requests after a time with no response because the initial request blocking the server.

Any help will be greatly appreciated

**Comment by carltongibson** [comment]
@JhonFrederick @ivanmviveros The immediate workaround here is to run the normal Django application with WSGI, and then only use ASGI for the async specific bits. 

You then route to either protocol server depending on URL say. 

Reviewing the underlying behaviour here is on my todo list but I won't have capacity to get to it instantly.

**Comment by Archmonger** [comment]
Confirmed to issue to persist on using this configuration:
`django==3.1.5`
`channels==3.0.3`
`daphne==3.0.1`

Additionally, confirmed workaround of routing http traffic through `wsgi.py` works. For onlookers, here's an example configuration to temporarily resolve this issue.
```python
# settings.py
ASGI_APPLICATION = "projectname.asgi.application"
WSGI_APPLICATION = "projectname.wsgi.application"
```
```python
# wsgi.py
from channels.routing import ProtocolTypeRouter
from django.core.wsgi import get_wsgi_application

application = ProtocolTypeRouter(
    {
        "http": get_wsgi_application(),
    }
)
```
```python
# asgi.py
import os
from django.conf.urls import url
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "conreq.settings")
from channels.auth import AuthMiddlewareStack
from channels.routing import ProtocolTypeRouter, URLRouter
from channels.security.websocket import AllowedHostsOriginValidator
from projectname.server_websockets import CommandConsumer

application = ProtocolTypeRouter(
    {
        "websocket": AllowedHostsOriginValidator(
            AuthMiddlewareStack(URLRouter([url("", CommandConsumer().as_asgi())]))
        ),
    }
)

```

**Comment by Archmonger** [comment]
Hey @carltongibson did you manage to trace down this issue to a particular part of the stack or LOC? 
If this is related to `asgrief` `thread_sensitive`, what was the Django team's reasoning behind using `thread_sensitive=true`?
I'd be willing to take a look at this and see what can be done.

**Comment by carltongibson** [comment]
Hey @Archmonger — Yes, this is about `thread_sensitive`. See the discussion here https://github.com/django/asgiref/issues/218 — there's a possible way forward. I haven't investigated in depth (for want of bandwidth). 

Very happy to input if you want to pick up to investigate though! 

Currently using WSGI for Django — tried and tested — and using ASGI only for Channels consumers (rather than running the whole stack under ASGI).

**Comment by Archmonger** [comment]
Got it, looks like I misunderstood the last comment on that asgrief thread.

I'll take a look at making the ORM async aware. I see an early comment about "subtle bugs". Is there an ongoing bug list for async ORM issues? For example, does Django suffer from half transmitted SQL queries (or intermingled queries) under these conditions?

**Comment by santuari** [comment]
> Additionally, confirmed workaround of routing http traffic through `wsgi.py` works as a temporary fix. For onlookers, here's an example configuration to temporarily resolve this issue.

As soon as I add channels in the `INSTALLED_APPS` all requests are handled sequentially both for `wsgi` and `asgi`. What I am doing wrong?

I use [your files](https://github.com/django/channels/issues/1587#issuecomment-780497550) and I run:

1. `daphne -b 127.0.0.1 -p 8082 my_app.asgi:application`\
2. `uwsgi --chdir=/my_app --module=my_app.wsgi:application --http=127.0.0.1:8080` or `gunicorn my_app.wsgi:application -b 127.0.0.1:8080`

Nginx is configured as follows:
```
server {
    listen 8001;

    location /ws/ {
        proxy_pass http://localhost:8082; # daphne (ASGI) listening on port 8082
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
    }

    location / {
        proxy_pass http://localhost:8081; # gunicorn (WSGI) listening on port 8081
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_connect_timeout 75s;
        proxy_read_timeout 300s;
        client_max_body_size 50m;
    }

    location /static {
      # path for static files
      root /my_app/;
    }
  }

  map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
  }
```

Versions:
```
asgiref  3.4.1
channels 3.0.4
Django 3.2.5
```

**Comment by Archmonger** [comment]
@santuari From what I've been told, this is a current limitation with channels (+asgiref). It had to be implemented to maintain thread safety while using channels. Fixing this is targeted for Django v4.

**Comment by santuari** [comment]
@Archmonger do you think that is possible to use channels + django in production? 
In our test environments, in which we have installed channels to support websocket with asgi deployment, two users are already impacted by very slow loading of view.

Anyway, I found a possible workaround by running gunicorn as async. 

Considering my previous comment,
>   uwsgi --chdir=/my_app --module=my_app.wsgi:application --http=127.0.0.1:8080 or gunicorn my_app.wsgi:application -b 127.0.0.1:8080

I used this command.
`gunicorn optimizer_app.wsgi:application -b 127.0.0.1:8080- --worker-class=gevent`

Requests are not blocked and daphne (point 1 of https://github.com/django/channels/issues/1587#issuecomment-884754933) can serve the "channels". 

I really don't know if this can be considered production-ready. What do you think?

**Comment by Archmonger** [comment]
@santuari Using two separate webserver and running WSGI+ASGI separately is the Django team's current workaround/suggestion for scalability while 4.0 is being developed.

If Websockets are not needed, then you can refactor your code to be fully WSGI while waiting for these updates.

**Comment by JhonFrederick** [comment]
At the development level, is the separation of the wsgi and the asgi also carried out? If so, does it affect debugging? In my case, I tried and I couldn't configure it in development and I chose to wait a solution of this issue. Any kind of help will be appreciated

**Comment by Archmonger** [comment]
@JhonFrederick See [my post above](https://github.com/django/channels/issues/1587#issuecomment-780497550) for info on separating WSGI and ASGI. 

Personally not using this workaround, but logically debugging shouldn't be too complicated using this. Your WSGI/ASGI stack traces would be separated in terminal unless you're running both servers through a single management command.

For my OSS projects, I've been working on a [ReactJS port to Django](https://github.com/idom-team/django-idom). Funneling everything through WS would mitigate this issue for me while the Django team works this ticket.

**Comment by jamiegau** [comment]
This problem hits my applications quite regularly when the channels WS gets busy specifically, it can end up locking up the django app for many minutes at a time for all users.  This is quite a serious issue and I have just been ignoring the problem for now as it's not happening that much..  Plus I expected some action on this problem by now. Personally I am super surprised its not getting more attention.  I imagine many other  web app makers typically use WS for real-time telemetry going to the users app.  Indicating whats going on with background tasks.  I do this for most of my apps.

If the above is the suggested production implementation, I would suggest the Channels manual updated to indicate as such. (https://channels.readthedocs.io/en/stable/deploying.html)

Django 4 is still quite a ways off, plus there is likely issues with it being a major release and compatibility with current apps.  I don't expect to be able to point at django 4 in requirements and start up the app and everything to work.

Otherwise, when I have time, after dealing with COVID crisis issues, I will implement this deployment workaround indicated above.

Thanks for the work on this by those involved.  I am greatly appreciative.  I am please this issues has been identified and a workaround and roadmap to overcome it is now formalised.

**Comment by mronoffon** [comment]
@jamiegau 

how many cores/workers are you using?

**Comment by jamiegau** [comment]
> @jamiegau
> 
> how many cores/workers are you using?

I always have at least 4 cores/workers in a VM running django.  Some times more depending on what its doing.

**Comment by PaleNeutron** [comment]
unbelievable the bug is still not fixed or a given a acceptable solution.

my env:

```
daphne==3.0.2
Django==3.2.7
```

Do something slow in one view will block all the other request.

**Comment by PaleNeutron** [comment]
> Hey @Archmonger — Yes, this is about `thread_sensitive`. See the discussion here [django/asgiref#218](https://github.com/django/asgiref/issues/218) — there's a possible way forward. I haven't investigated in depth (for want of bandwidth).
> 
> Very happy to input if you want to pick up to investigate though!
> 
> Currently using WSGI for Django — tried and tested — and using ASGI only for Channels consumers (rather than running the whole stack under ASGI).

So how to use `async_to_sync(layer.group_send)` in this solution?

I need `layer.group_send` in a http view. If the view is served by `wsgi` app, that won't work.

**Comment by revoteon** [comment]
This is had immense impact on our app performance. For those who are impacted, the suggested workaround of serving sync and async separately works. You _will_ notice significant performance gain with this workaround. 

On a separate note, I believe the workaround should be included in the docs, especially if a permanent solution is not in the horizon.

**Comment by revoteon** [comment]
@PaleNeutron What WSGI server are you running? In your case I'd try to run `gunicorn` in `gthread` mode with `--threads` set to vCPU count + 1.

**Comment by PaleNeutron** [comment]
@revoteon ,I use  uwsgi to serve wsgi app as mentioned above

> 
> As soon as I add channels in the INSTALLED_APPS all requests are handled sequentially both for wsgi and asgi. What I am doing wrong?
> 
> I use your files and I run:
> 
> daphne -b 127.0.0.1 -p 8082 my_app.asgi:application\
> uwsgi --chdir=/my_app --module=my_app.wsgi:application --http=127.0.0.1:8080 or gunicorn my_app.wsgi:application -b 127.0.0.1:8080
> Nginx is configured as follows:
>

**Comment by revoteon** [comment]
@PaleNeutron By [default](https://uwsgi-docs.readthedocs.io/en/latest/WSGIquickstart.html#adding-concurrency-and-monitoring) `uwsgi` start with single process and single thread. You should add `--processes` and `--threads` arguments to add concurrency. Values for processes and threads need some fine tuning, depending on your deployment environment, system resources and etc. But `2` for processes and (vCPU count + 1) for threads is a good start.

All these apply to `gunicorn` as well, with gthread workers.

**Comment by MarcoRizk** [comment]
In case anyone is stuck with this issue I wrote [here](https://www.linkedin.com/pulse/django-3-channels-bad-recipe-what-can-you-do-marco-rizk/) how to solve it by separating wsgi / asgi apps. Using websockets url prefix and routing to the asgi app through Nginx. And setting up the entire thing in docker-compose with example code. Hope you find it helpful also enhancments to my implementation are welcome :)
https://www.linkedin.com/pulse/django-3-channels-bad-recipe-what-can-you-do-marco-rizk/

**Comment by carltongibson** [comment]
Nice write up @MarcoRizk. Summarises it well. 👍

**Comment by revoteon** [comment]
@MarcoRizk Good write up 👍

I'd add one caveat in the doc about gunicorn. By default gunicorn runs a sync worker, that is each worker can handle 1 request at a time. In this default mode, you have to spin up multiple workers to enable concurrency. In your document you spin up 4, which means the app will handle at most 4 requests simultaneously. It also means it'll take 4 times the memory.

You may instead use `gthread` worker and specify multiple threads (and optionally workers) to achieve much higher throughput.  This can be implicitly enabled by `--threads` argument.

How many threads and workers you should specify depends on your app, infrastructure/resources and connection capacity of your db. It is a balancing act. But count(cpu) + 1 for threads is a good start. With this in mind, for example, gunicorn with 4 workers and 4 threads each will be able handle 16 requests concurrently. 

Unfortunately Django docs doesn't mention this aspect of gunicorn (or uwsgi) deployment, which I believe is very crucial.

Additionally in container based deployments, like kubernetes, you may choose to specify 1 worker and multiple threads per pod and scale pods horizontally as per your metrics.

**Comment by bj00rn** [comment]
Does this issue affect channel 3 running django 2 as well? or is this django 3 only?

**Comment by carltongibson** [comment]
@bj00rn This affects the `ASGIHander` wrapper of a traditional Django-app, whether that has an async pathway (Django 3.1+ or not). 

Django 4.0 introduces per-request async contexts, which allows a thread per-request here, which should see a good improvement. (The ORM remains fully-sync, so we're still limited to a single thread per-request for DB operations, for the now.) 

Likely, the best approach here for the now is to serve your main Django app using WSGI (with gunicorn &co) as always, and route a separate ASGI app just for the async parts, be they Channels or async Django views.

**Comment by Archmonger** [comment]
"Django 4.0 introduces per-request async contexts"

Is this only when using `sync_to_async` with `thread_sensitive=False`?

**Comment by carltongibson** [comment]
@Archmonger No, the thread sensitivity is per-request. (That's the advantage of the change.) 

(It's one single-threaded ThreadPoolExecuter per request, rather than one for all requests.)

**Comment by keithhackbarth** [comment]
To answer my own question, need to follow the deprecation notice here:
https://github.com/django/channels/blob/e6eb6f7c2a0ed8c2829e9368878b694ba226c675/channels/routing.py#L40

And if you use StaticFileHandler 
https://github.com/django/channels/pull/1800/files

**Comment by jeloagnasin** [comment]
@keithhackbarth, sorry, don't quite get what you mean. Are you confirming that Django 4 resolves this issue?

**Comment by keithhackbarth** [comment]
Yes. However if you are upgrading, make sure to update your routing according to the depreciation message. The default fallback ASGIHandler in channels still has bug.

**Comment by jeloagnasin** [comment]
> Yes. However if you are upgrading, make sure to update your routing according to the depreciation message. The default fallback ASGIHandler in channels still has bug.

Awesome. Thanks! 😁

**Comment by Archmonger** [comment]
I can confirm this as well.

Artificially putting a `sleep(10)` in one view does not prevent the rendering of other views.

**Comment by carltongibson** [comment]
This is effectively resolved by upgrading to Django 4.0.

**Comment by chrisbarry** [comment]
Can we get this put into some docs somewhere, either on django or channels? (Perhaps it is) but as I had decided only to go up to django 3, with channels 2 and I was running via gunicorn with uvicorn inside it, I got caught out. Anyway I could have caught this earlier (aside from performance testing on staging more agressively) I stupidly assuming nothing this major could happen as part of the upgrade process...
